### DISCUSSION TITLE
Evaluate Profile-Guided Optimization (PGO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [vlcn-io/cr-sqlite#403] [Ideas]
Hi.

There are a lot of examples of different software, where Profile-Guided Optimization (PGO) helps a lot with performance - you can check it [here](https://github.com/zamazan4ik/awesome-pgo). E.g. in this list are a lot of databases like Redis, MongoDB, and PostgreSQL. For CR-SQLite the most interesting results I think the PGO [results](https://github.com/zamazan4ik/awesome-pgo/blob/main/sqlite.md) for SQLite.

We need to check the PGO effects on `cr-sqlite`, and if it works well - write a note in the `cr-sqlite` documentation about PGO (like it's already [done](https://clickhouse.com/docs/en/operations/optimizing-performance/profile-guided-optimization) for ClickHouse). I'd appreciate your providing an easy way to build `cr-sqlite` with PGO, and experienced users will be able to do it on their own for their own usage scenarios. Another option is to optimize the `cr-sqlite` builds with a generic-enough profile. Providing PGO-optimized binaries could be a trickier task (since it requires preparing a good-enough profile) but as an option would be great to see too. 

For the Rust projects, I recommend starting with [cargo-pgo](https://github.com/Kobzol/cargo-pgo).

**Comment by tantaman** [comment]
maybe one day... not quite a priority for me right now.

**Comment by zamazan4ik** [comment]
> maybe one day... not quite a priority for me right now.

Maybe will be better to leave the issue open? In this case, maybe someone from the community will take the issue and perform all corresponding benchmarks and other PGO-related stuff.

**Comment by tantaman** [comment]
Sure, will do.
